File: src/dataStructures/singlyLinkedList/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
 var addTwoNumbers = function(l1, l2) {
    let digits;
    let l3 = new ListNode();
    let surplus = 0;
    let curr = l3;
    while(l1 || l2){
        digits = (l1?.val || 0) + (l2?.val || 0) + surplus;
        if(l1) l1= l1.next;
        if(l2) l2= l2.next;
        digits = digits + "";
        if(digits.length ===2){
            surplus = digits[0];
            surplus = parseInt(surplus);
            digits = digits[1];
        } else {
            surplus = 0;
        }
        curr.val = parseInt(digits);
        if(l1 || l2) {
            curr.next = new ListNode();
            curr = curr.next;
        } else if(surplus) curr.next = new ListNode(surplus);
    }
    return l3;
};

File: src/dataStructures/singlyLinkedList/addTwoNumbers.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    list1 = l1
    list2 = l2
    prehead = ListNode.new(0);
    curr = prehead
    carry_over = 0
    while list1 || list2
        sum = ((list1 && list1.val) || 0) + ((list2 && list2.val) || 0) + carry_over
        list1 = list1 ? list1.next : nil
        list2 = list2 ? list2.next : nil
        curr.next = ListNode.new(sum%10)
        curr = curr.next
        carry_over = ((sum/10)%10).floor
    end
    carry_over > 0 ? curr.next = ListNode.new(carry_over) : nil
    prehead.next
end


l1 = ListNode.new(2)
l1.next = ListNode.new(4)
l1.next.next = ListNode.new(3)
l2 = ListNode.new(5)
l2.next = ListNode.new(6)
l2.next.next = ListNode.new(4)

add_two_numbers(l1,l2)
